src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -312,6 +312,33 @@ impl ScalarFunc {
             _ => Err(box_err!("invalid direction value: {}", direction)),
         }
     }
+
+    #[inline]
+    pub fn substring_index<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let delim = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
+        let delim = delim.as_ref();
+        let count = try_opt!(self.children[2].eval_int(ctx, row));
+
+        let strs = s.as_ref().split(delim).collect::<Vec<_>>();
```

Comment:
There could be a more efficient way instead of collecting all splitted strs.

Reply:
In my first implemention , I used iter API 
when `count > 0` ,  we can only collect only the needed strs and join
```
s.split(delim).take(count).collect::<Vec<_>>().join(delim);
```
when `count < 0`,  we need the length to calculate the skip_number
```
let length = s.as_ref().split(delim).count();
 let r = s
        .as_ref()
        .split(delim)
        .skip(length - length.min(-count as usize))
        .collect::<Vec<_>>()
        .join(delim);
```
I'm not quite sure is this efficient, since  it will construct and  consume `std::str::Split` two times